pandas/io/parquet.py

```diff
@@ -302,16 +305,27 @@ def to_parquet(
     if isinstance(partition_cols, str):
         partition_cols = [partition_cols]
     impl = get_engine(engine)
-    return impl.write(
+
+    if path is None:
+        path_or_buf = io.BytesIO()
+    else:
+        path_or_buf = path
+
+    impl.write(
         df,
-        path,
+        path_or_buf,
         compression=compression,
         index=index,
         partition_cols=partition_cols,
         storage_options=storage_options,
         **kwargs,
     )
 
+    if path is None:
+        return path_or_buf.getvalue()
```

Comment:
Getting a mypy error here:
```
pandas/io/parquet.py:312: error: Incompatible types in assignment (expression has type "Union[str, Path, IO[Any], IOBase]", variable has type "BytesIO")  [assignment]
```
which doesn't make sense because `BytesIO` subclasses `IOBase`

Reply:
you need to assert that path_or_buf is bytes i think.